File: clients/protocol-client/src/types/ruleset/terrain_gen.ts
import { ShadyRegister, ShasmProgram, ShasmProgramValidation } from "../shady_vm";
import { FormatRules, FormatInput, FormatValidation } from "./format_rules";

export type TerrainGenRules = {
  perlin: TerrainGenPerlinRules,
  stage: TerrainGenStageRules,
};
export type TerrainGenInput = {
  perlin: TerrainGenPerlinInput,
  stage: TerrainGenStageInput,
};
export type TerrainGenValidation = {
  errors: string[],
  perlin: TerrainGenPerlinValidation,
  stage: TerrainGenStageValidation,
};
export function defaultTerrainGenRules(): TerrainGenRules {
  return {
    perlin: defaultTerrainGenPerlinRules(),
    stage: defaultTerrainGenStageRules(),
  };
}
export function validateTerrainGenRules(
  rules: TerrainGenRules,
  errors: string[]
): boolean {
  // Validate perlin rules
  validateInteger(rules.perlin.octaves, 1, 9, "Perlin octaves", errors);
  validateInteger(rules.perlin.frequency, 1, 99, "Perlin frequency", errors);
  validateInteger(rules.perlin.amplitude, 1, 99, "Perlin amplitude", errors);

  // Validate stage rules
  validateInteger(rules.stage.iterations, 1, 9999, "Stage iterations", errors);

  // Validate format rules
  validateInteger(rules.stage.format.wordFormats.length, 1, 63,
    "Stage format word count", errors);

  return errors.length === 0;
}

function validateInteger(
  value: number,
  min: number,
  max: number,
  name: string,
  errors: string[]
): void {
  if (!Number.isInteger(value)) {
    errors.push(`${name} must be an integer`);
  }
  if (value < min) {
    errors.push(`${name} must be at least ${min}`);
  }
  if (value > max) {
    errors.push(`${name} must be at most ${max}`);
  }
}

export type TerrainGenStageRules = {
  format: FormatRules,
  initProgram: ShasmProgram,
  iterations: number,
  pairwiseProgram: ShasmProgram,
  pairwiseOutputRegisters: number,
  mergeProgram: ShasmProgram,
  finalProgram: ShasmProgram,
};
export type TerrainGenStageInput = {
  format: FormatInput,
  initProgram: string,
  iterations: string,
  pairwiseProgram: string,
  pairwiseOutputRegisters: string,
  mergeProgram: string,
  finalProgram: string,
};
export type TerrainGenStageValidation = {
  errors: string[],

  format: FormatValidation,
  initProgram: ShasmProgramValidation,
  iterations: string[],
  pairwiseProgram: ShasmProgramValidation,
  pairwiseOutputRegisters: string[],
  mergeProgram: ShasmProgramValidation,
  finalProgram: ShasmProgramValidation,
};
function defaultTerrainGenStageRules(): TerrainGenStageRules {
  return {
    format: { wordFormats: [] },
    initProgram: { programText: "" },
    iterations: 1,
    pairwiseProgram: { programText: "" },
    pairwiseOutputRegisters: 0,
    mergeProgram: { programText: "" },
    finalProgram: { programText: "" },
  };
}

export type TerrainGenPerlinRules = {
  seed: number,
  octaves: number,
  frequency: number,
  amplitude: number,

  register: ShadyRegister,
};
export type TerrainGenPerlinInput = {
  seed: string,
  octaves: string,
  frequency: string,
  amplitude: string,

  register: string,
};
export type TerrainGenPerlinValidation = {
  errors: string[],

  seed: string[],
  octaves: string[],
  frequency: string[],
  amplitude: string[],

  register: string[],
};
function defaultTerrainGenPerlinRules(): TerrainGenPerlinRules {
  return {
    seed: 1,
    octaves: 1,
    frequency: 1,
    amplitude: 1,
    register: 0,
  };
}

File: src/data/ruleset/terrain_gen.rs
use crate::shady_vm::{ ShadyRegister, ShasmProgram };
use super::{
  terrain_gen_randgen::{
    TerrainGenPerlinRules,
    TerrainGenPerlinInput,
    TerrainGenPerlinValidation,
  },
  terrain_gen_stage::{
    TerrainGenStageRules,
    TerrainGenStageInput,
    TerrainGenStageValidation,
  },
  FormatRules
};


#[derive(Debug, Clone)]
#[derive(serde::Serialize, serde::Deserialize)]
pub(crate) struct TerrainGenRules {
  // The terrain generation starts with some number of optional perlin passes
  // to initialize one or more registers of the terrain generator VM.
  pub(crate) perlin: TerrainGenPerlinRules,

  // The definition of each terrain generator pass.
  pub(crate) stage: TerrainGenStageRules,
}
impl TerrainGenRules {
  pub(crate) fn to_input(&self) -> TerrainGenInput {
    TerrainGenInput {
      perlin: self.perlin.to_input(),
      stage: self.stage.to_input(),
    }
  }
}
impl TerrainGenRules {
  pub(crate) fn new_example() -> Self {
    TerrainGenRules {
      perlin: TerrainGenPerlinRules {
        register: ShadyRegister::new(92),
      },
      stage: TerrainGenStageRules {
        format: FormatRules::new_example(),
        init_program: ShasmProgram::new_example(),
        pairwise_program: ShasmProgram::new_example(),
        merge_program: ShasmProgram::new_example(),
        final_program: ShasmProgram::new_example(),
      },
    }
  }
}

#[derive(Debug, Clone)]
#[derive(serde::Serialize, serde::Deserialize)]
pub(crate) struct TerrainGenInput {
  pub(crate) perlin: TerrainGenPerlinInput,
  pub(crate) stage: TerrainGenStageInput,
}
impl TerrainGenInput {
  pub(crate) fn new() -> Self {
    TerrainGenInput {
      perlin: TerrainGenPerlinInput::new(),
      stage: TerrainGenStageInput::new(),
    }
  }

  pub(crate) fn to_validated(&self) -> Result<TerrainGenRules, TerrainGenValidation> {
    let perlin = self.perlin.to_validated();
    let stage = self.stage.to_validated();
    if perlin.is_ok() && stage.is_ok() {
      Ok(TerrainGenRules {
        perlin: perlin.unwrap(),
        stage: stage.unwrap()
      })
    } else {
      Err(TerrainGenValidation {
        errors: vec!["The terrain generator is invalid.".to_string()],
        perlin: perlin.err(),
        stage: stage.err(),
      })
    }
  }
}

#[derive(Debug, Clone)]
#[derive(serde::Serialize, serde::Deserialize)]
pub(crate) struct TerrainGenValidation {
  pub(crate) errors: Vec<String>,

  #[serde(skip_serializing_if = "Option::is_none")]
  pub(crate) perlin: Option<TerrainGenPerlinValidation>,

  #[serde(skip_serializing_if = "Option::is_none")]
  pub(crate) stage: Option<TerrainGenStageValidation>,
}
impl TerrainGenValidation {
  pub(crate) fn new() -> Self {
    TerrainGenValidation {
      errors: Vec::new(),
      perlin: None,
      stage: None,
    }
  }

  pub(crate) fn is_valid(&self) -> bool {
    self.errors.is_empty()
      && self.perlin.as_ref().map_or(true, |pv| pv.is_valid())
      && self.stage.as_ref().map_or(true, |sv| sv.is_valid())
  }
}
